Feat / Hero and HeroShelf improvements

royschut opened the pull request:
## Description

<!-- Describe the proposed solution/fix/feature in this pull request here. -->

This PR improves the Hero and HeroShelf and removes Card shadow:
- Fade out the background incrementally based on scroll position. This brings a much smoother scrolling experience.
- Add swiping gesture to HeroShelf for all breakpoints. Previously swiping was only possible for mobile devices, but on tablets or landscape mobile it is needed as well to navigate between slides. I have tried simplifying the code by merging the two implementations, leaving only style differences between mobile and the rest.
- Some minor style updates to HeroShelf: a bit more height on tablet, better support for mobile landscape and respecting the safe-area better (useful for non-web platforms).
- Make the Hero fixed as well, and add the same fade behaviour.
- Remove the Card box-shadow. Since we've re-introduced support for a light background, the card shadow seemed a bit overdone, or out-dated even in some opinions (have a look here: https://vd-ott.com/?app-config=kcm2oevm). We figured that the card shadow could be left out entirely, as it was once introduced mainly because of the blur background, which was removed quite some time ago. 

### Steps completed:

<!-- Check all completed steps so we know  -->

According to our definition of done, I have completed the following steps:

- [x] Acceptance criteria met
- [ ] Unit tests added
- [ ] Docs updated (including config and env variables)
- [ ] Translations added
- [x] UX tested
- [x] Browsers / platforms tested
- [x] Rebased & ready to merge without conflicts
- [x] Reviewed own code

langemike commented:
These changes add a lot of value! Swiping works better, indeed. I think we're on the right track, but I noticed some issues that are related to different (smartphone) resolutions.

There are 3 styling-related issues. These weren't there previously.
1. The Hero title can get cut-off on top. 
2. The buttons look crammed now (too narrow button container)

3. Also in some specific resolutions, the Start Watching button doesn't show, as can be seen here:
![Screenshot 2024-11-14 at 13 18 07](https://github.com/user-attachments/assets/114c785a-366b-4dc1-8405-4e3258a2c46a)

And I also noticed 2 issues in the scrolling (fade to transparent) behaviour:
1. It should finish its transition when the hero is out of view. But it happens around the second regular shelf
2. It doesn't fade completely away

These happen on certain resolutions such as 2340x1080 (on my phone). On other resolution, I see it works as intended.

I tested this on the deploy preview.

Video recording showcasing styling issues 1 & 2 and scrolling issues 1 & 2.

https://github.com/user-attachments/assets/4f9bc07f-54d0-4a9b-a31f-3a3f9cc12c30

ChristiaanScheermeijer commented on packages/ui-react/src/components/HeroShelf/HeroShelfMetadata.tsx:
```diff
@@ -21,17 +22,18 @@ const HeroShelfMetadata = ({
   loading: boolean;
   playlistId: string | undefined;
-  style?: CSSProperties;
+  style: CSSProperties;
```
Is this an unwanted change? I do believe `isMobile?: boolean;` can be removed

royschut replied:
It's intentional, style should now always be set.
isMobile is removed in my last commit.

ChristiaanScheermeijer commented on platforms/web/src/styles/main.scss:
```diff
@@ -19,10 +19,13 @@
-  --safe-area-right: 0;
-  --safe-area-bottom: 0;
-  --safe-area-left: 0;
+  /* stylelint-disable length-zero-no-unit */
+  --safe-area-top: 0px; 
+  --safe-area-right: 0px;
+  --safe-area-bottom: 0px;
+  --safe-area-left: 0px;
+  /* stylelint-enable length-zero-no-unit */
```
How come using `0` is a problem?

royschut replied:
Chrome doesn't respect it within a `calc` function, and ignores setting the shelf height.

ChristiaanScheermeijer commented on packages/ui-react/src/components/HeroShelf/HeroShelf.module.scss:
```diff
@@ -97,6 +98,7 @@ $mobile-landscape-height: 100vh;
   height: 56.25vw;
   max-height: 700px;
   background-color: var(--hero-shelf-background-color);
+  transition: opacity ease-out 0.3s;
```
IMO the scroll animation feels better without using an extra CSS transition (no delay). The scroll events already defines the animation timeline and also is more performance friendly, but that is probably not even noticeable.

royschut replied:
Thanks, I agree. Changed!

ChristiaanScheermeijer commented on packages/ui-react/src/components/HeroShelf/HeroShelf.tsx:
```diff
@@ -27,20 +27,28 @@ const HeroShelf = ({ playlist, loading = false, error = null }: Props) => {
 
-  const slideTo = (toIndex: number) => {
+  useScrolledDown(50, isMobile ? 200 : 700, (progress: number) => {
+    if (posterRef.current) posterRef.current.style.opacity = `${Math.max(1 - progress, isMobile ? 0 : 0.1)}`;
```
Should we also fade away the pagination?

royschut replied:
I didn't find it necessary anymore. Before they were also fixed, but they scroll away now.

ChristiaanScheermeijer commented on packages/ui-react/src/components/HeroShelf/HeroSwipeSlider.tsx:
```diff
@@ -73,18 +88,17 @@ const HeroShelfMetadataMobile = ({ item, leftItem, rightItem, playlistId, loadin
       containerRef.current.style.transition = 'transform 0.2s ease-out';
       containerRef.current.style.transform = `translateX(100%)`;
-    } else if (direction === 'right') {
+    } else if (isSwipeAnimation && direction === 'right') {
```
Would it look nicer for tablets and larger screen sizes to decrease the left/right offset and use opacity as well? It works great now, but the slide animation distance is pretty big (opinionated ofc)

royschut replied:
I think you mean the animation after a swipe gesture (because the normal slide hasn't changed). The thing is: when you drag the metadata over the screen, then let go, you want the gesture to be continued similarly. So I cannot really change the opacity, you expect it to be opaque. Same for the offset, it comes from outside of the screen. I could only update the transition perhaps?

royschut replied:
We can consider this later as well!

royschut commented:
> There are 3 styling-related issues. These weren't there previously.
> 
> 1.

Thanks for the thorough testing @langemike! My last commit fixes the issues you mentioned.
About the scrolling fade: I have updated the end of the scroll/opacity window slightly, but if I change it too much, I find the transition a bit too sudden. You're now previewing on landscape, which is quite a specific viewport, I prefer to focus on more used viewports (without introducing too many variations). I have also removed the transition, as Christiaan suggested, which might help in your experience. How does the latest update feel?
Leaving 0.1 opacity is a choice by the way.

langemike commented:
> How does the latest update feel?

Better but the video details section does a weird jump now during the slide animation.

> Leaving 0.1 opacity is a choice by the way.

Ok. Good to know!

I think the fade out to 0.1 still happens quite late. But I can live with that.

updated video recording:
https://github.com/user-attachments/assets/df89d9a8-8bd5-40a4-86ed-44551699325b

ps. This can also be reproduced in Chrome using "responsive mode" using 2340x1080 or compatible screen size.

royschut commented:
> I think the fade out to 0.1 still happens quite late. But I can live with that.

It's a compromise on mobile landscape, to bring the best experience on more common viewports (mobile portrait, tablet and desktop). The only alternative I see is introducing more variation for the scroll/fade position, but since it is done from javascript, I prefer to keep it simple. But open to suggestions.
